osquery/tables/system/linux/deb_packages.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include <boost/algorithm/string.hpp>
+#include <osquery/tables.h>
+
+// see README.api of libdpkg-dev
+#define LIBDPKG_VOLATILE_API
+
+extern "C" {
+#include <dpkg/dpkg-db.h>
+// copy pasted from dpkg-db.h
+// these enums are inside struct pkginfo and are not visible for other headers
+enum pkgwant {
+  want_unknown,
+  want_install,
+  want_hold,
+  want_deinstall,
+  want_purge,
+  /** Not allowed except as special sentinel value in some places. */
+  want_sentinel,
+} want;
+/** The error flag bitmask. */
+enum pkgeflag {
+  eflag_ok = 0,
+  eflag_reinstreq = 1,
+} eflag;
+enum pkgstatus {
+  stat_notinstalled,
+  stat_configfiles,
+  stat_halfinstalled,
+  stat_unpacked,
+  stat_halfconfigured,
+  stat_triggersawaited,
+  stat_triggerspending,
+  stat_installed
+} status;
+
+#include <dpkg/dpkg.h>
+#include <dpkg/pkg-array.h>
+#include <dpkg/parsedump.h>
+}
+
+namespace osquery {
+namespace tables {
+
+// dpkg versions in 12.04 and 14.04 use completely different sorters
+// bringing this one forward to smoother over the incompatibility
+int pkg_sorter(const void *a, const void *b) {
+  const struct pkginfo *pa = *(const struct pkginfo **)a;
+  const struct pkginfo *pb = *(const struct pkginfo **)b;
+  const char *arch_a = pa->installed.arch->name;
+  const char *arch_b = pb->installed.arch->name;
+  int res;
+
+  res = strcmp(pa->set->name, pb->set->name);
+  if (res)
+    return res;
+
+  if (pa->installed.arch == pb->installed.arch)
+    return 0;
+
+  return strcmp(arch_a, arch_b);
+}
+
+void dpkg_setup(char const *name,
+                enum modstatdb_rw mode,
+                struct pkg_array *array) {
+  dpkg_set_progname(name);
+  push_error_context();
+
+  modstatdb_open(mode);
+
+  pkg_array_init_from_db(array);
+  pkg_array_sort(array, pkg_sorter);
+}
+
+void dpkg_teardown(struct pkg_array *array) {
+  pkg_array_destroy(array);
+
+  pkg_db_reset();
+  modstatdb_done();
+
+  pop_error_context(ehflag_normaltidy);
+}
+
+const std::map<std::string, std::string> kFieldMappings = {
+    {"Package", "name"},
+    {"Version", "version"},
+    {"Installed-Size", "size"},
+    {"Architecture", "arch"},
+    {"Source", "source"},
+    {"MD5sum", "md5sum"}, };
+
+/**
+* Fields information - taken from lib/dpkg/parse.c
+*/
+const struct fieldinfo fieldinfos[] = {
+    /* Note: Capitalization of field name strings is important. */
```

Comment:
Can we change these to C++ style comments. The doxygen generation comment blocks above methods/data structures are C style though. :) 

Also, just curious, do you need to enumerate all the fields, can we limit to just the fields that are emitted in the table spec? Or do you recommend keeping them all here for completeness/future additions.